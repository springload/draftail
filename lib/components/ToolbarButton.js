import PropTypes from "prop-types";
import React, { PureComponent } from "react";

import Icon from "./Icon";

/**
 * Displays a basic button, with optional active variant,
 * enriched with a tooltip. The tooltip stops showing on click.
 */
class ToolbarButton extends PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      showTooltipOnHover: true,
    };

    this.onMouseDown = this.onMouseDown.bind(this);
    this.onMouseLeave = this.onMouseLeave.bind(this);
  }

  onMouseDown(e) {
    const { name, onClick } = this.props;

    e.preventDefault();

    this.setState({
      showTooltipOnHover: false,
    });

    onClick(name);
  }

  onMouseLeave() {
    this.setState({
      showTooltipOnHover: true,
    });
  }

  render() {
    const { name, active, label, title, icon } = this.props;
    const { showTooltipOnHover } = this.state;

    return (
      <button
        name={name}
        className={`Draftail-ToolbarButton${
          active ? " Draftail-ToolbarButton--active" : ""
        }`}
        type="button"
        aria-label={title || null}
        data-draftail-balloon={title && showTooltipOnHover ? true : null}
        tabIndex={-1}
        onMouseDown={this.onMouseDown}
        onMouseLeave={this.onMouseLeave}
      >
        {icon ? <Icon icon={icon} /> : null}
        {label ? (
          <span className="Draftail-ToolbarButton__label">{label}</span>
        ) : null}
      </button>
    );
  }
}

ToolbarButton.propTypes = {
  name: PropTypes.string,
  active: PropTypes.bool,
  label: PropTypes.string,
  title: PropTypes.string,
  icon: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string),
    PropTypes.node,
  ]),
  onClick: PropTypes.func,
};

ToolbarButton.defaultProps = {
  name: null,
  active: false,
  label: null,
  title: null,
  icon: null,
  onClick: () => {},
};

export default ToolbarButton;
