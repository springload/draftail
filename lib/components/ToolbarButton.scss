$button-active-color: transparentize(
  $color: $draftail-editor-chrome-active,
  $amount: 0.9,
);
$button-active-border-color: transparentize(
  $color: $draftail-editor-chrome-active,
  $amount: 0.8,
);
$button-font-size: 1rem;

.#{$DRAFTAIL}ToolbarButton {
  @include font-smoothing();

  display: inline-block;
  padding: $button-spacing;
  border-radius: 0;
  background: transparent;
  color: inherit;
  border: 1px solid transparent;
  font-size: $button-font-size;
  font-weight: 600;
  cursor: pointer;
  user-select: none;

  &--active {
    background-color: $button-active-color;
    border: 1px solid $button-active-border-color;
  }

  &__label {
    display: inline-block;
    vertical-align: middle;
    height: 1em;
  }
}
