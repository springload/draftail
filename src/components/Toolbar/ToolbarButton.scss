@use "sass:color";
@use "../../api/constants" as *;

$button-active-color: color.adjust(
  $color: $draftail-editor-chrome-active,
  $alpha: -0.9,
);
$button-active-border-color: color.adjust(
  $color: $draftail-editor-chrome-active,
  $alpha: -0.8,
);
$button-font-size: 1rem;

.#{$DRAFTAIL}ToolbarButton {
  // Guarantee buttons are a predictable round-number size.
  $size: $button-font-size + $button-spacing * 2 + $draftail-base-spacing;
  @include font-smoothing();

  display: inline-block;
  padding: $button-spacing;
  min-width: $size;
  max-height: $size;
  border-radius: $draftail-toolbar-radius;
  background: transparent;
  color: inherit;
  border: 1px solid transparent;
  outline: $draftail-contrast-outline;
  font-size: $button-font-size;
  font-weight: 600;
  cursor: pointer;
  user-select: none;

  &--active {
    background-color: $button-active-color;
    border: 1px solid $button-active-border-color;

    @media (forced-colors: active) {
      // Force the text to CanvasText as forced colors automatically
      // adds a Canvas outline behind text.
      background: Highlight;
      color: HighlightText;
      border-color: currentColor;
    }
  }

  &__label {
    display: inline-block;
    vertical-align: middle;
    height: 1em;

    @media (forced-colors: active) {
      color: CanvasText;
    }
  }
}
