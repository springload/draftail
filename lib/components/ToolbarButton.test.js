import React from "react";
import { shallow } from "enzyme";
import ToolbarButton from "./ToolbarButton";

describe("ToolbarButton", () => {
  it("empty", () => {
    expect(shallow(<ToolbarButton />)).toMatchSnapshot();
  });

  it("#name", () => {
    expect(shallow(<ToolbarButton name="TEST" />)).toMatchSnapshot();
  });

  it("#label", () => {
    expect(shallow(<ToolbarButton label="Test label" />)).toMatchSnapshot();
  });

  it("#icon", () => {
    expect(shallow(<ToolbarButton icon="icon-test" />)).toMatchSnapshot();
  });

  it("#active", () => {
    expect(shallow(<ToolbarButton active />)).toMatchSnapshot();
  });

  it("#title", () => {
    expect(shallow(<ToolbarButton title="Test title" />)).toMatchSnapshot();
  });

  it("#onClick", () => {
    const onClick = jest.fn();
    const event = { preventDefault: jest.fn() };
    shallow(<ToolbarButton onClick={onClick} />).simulate("mousedown", event);
    expect(onClick).toHaveBeenCalledTimes(1);
    expect(event.preventDefault).toHaveBeenCalledTimes(1);
    expect();
  });

  it("stops showing tooltip on click", () => {
    const wrapper = shallow(<ToolbarButton />);
    expect(wrapper.state("showTooltipOnHover")).toBe(true);

    wrapper.simulate("mousedown", {
      preventDefault: jest.fn(),
    });
    expect(wrapper.state("showTooltipOnHover")).toBe(false);

    wrapper.simulate("mouseleave", null);
    expect(wrapper.state("showTooltipOnHover")).toBe(true);
  });
});
