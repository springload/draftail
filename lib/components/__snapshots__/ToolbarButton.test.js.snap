// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolbarButton #active 1`] = `
<button
  aria-label={null}
  className="Draftail-ToolbarButton Draftail-ToolbarButton--active"
  data-draftail-balloon={null}
  name={null}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  tabIndex={-1}
  type="button"
/>
`;

exports[`ToolbarButton #icon 1`] = `
<button
  aria-label={null}
  className="Draftail-ToolbarButton"
  data-draftail-balloon={null}
  name={null}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  tabIndex={-1}
  type="button"
>
  <Icon
    className={null}
    icon="icon-test"
    title={null}
  />
</button>
`;

exports[`ToolbarButton #label 1`] = `
<button
  aria-label={null}
  className="Draftail-ToolbarButton"
  data-draftail-balloon={null}
  name={null}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  tabIndex={-1}
  type="button"
>
  Test label
</button>
`;

exports[`ToolbarButton #name 1`] = `
<button
  aria-label={null}
  className="Draftail-ToolbarButton"
  data-draftail-balloon={null}
  name="TEST"
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  tabIndex={-1}
  type="button"
/>
`;

exports[`ToolbarButton #title 1`] = `
<button
  aria-label="Test title"
  className="Draftail-ToolbarButton"
  data-draftail-balloon={true}
  name={null}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  tabIndex={-1}
  type="button"
/>
`;

exports[`ToolbarButton empty 1`] = `
<button
  aria-label={null}
  className="Draftail-ToolbarButton"
  data-draftail-balloon={null}
  name={null}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  tabIndex={-1}
  type="button"
/>
`;
